lib/addMethod/addMethodSOAP.js

```diff
@@ -0,0 +1,207 @@
+var when               = require('when');
+var _                  = require('lodash');
+var setParam           = require('mout/queryString/setParam');
+var guid               = require('mout/random/guid');
+var logger             = require('../logger');
+var globalize          = require('./globalize');
+var substitute         = require('./substitute');
+
+var validateInput      = require('./validateSOAPInput');
+
+var validateExpects    = require('./validateExpects');
+var validateNotExpects = require('./validateNotExpects');
+
+var getSOAPClient = require('./getSOAPClient');
+
+
+module.exports = function (methodName, config) {
+
+	var threadneedle = this;
+
+	validateInput.call(this, methodName, config);
+
+	threadneedle[methodName] = function (params) {
+
+		params = params || {};
+
+		var globalsDisabled = ( !_.isUndefined(config.globals) && config.globals === false );
+
+		if (_.isFunction(config)) {
+	      logger.info(methodName+': Running method function.');
+	      return when(config.call(threadneedle, params));
+	    }
+
+		logger.info('Running method `' + methodName + '` `before`.');
+
+
+		//Do SOAP stuff
+		return when.promise(function (resolve, reject) {
+
+			// Kick off that promise chain
+			when()
+
+			// Run a `before` if set on the params.
+			.then(function () {
+				return globalize.before.call(threadneedle, config, params);
+			})
+
+			// Set a bunch of local variables, formatted and templated
+			.then(function (result) {
+
+				params = result;
+
+				logger.info(methodName+': substituting parameters');
+
+				//Process all the required data prior to executing the actual request
+				var fullMethodPath = globalize.string.call(threadneedle, config, params, 'method', 'baseMethod');
```

Comment:
Forgive the silly question - but why is this needed? methods in SOAP are just functions from my understanding, why would you want to declare some of it on the global level and some locally?

Reply:
It's there if it's needed. 
E.g. if all the method names have a common prefix, the prefix part can go in the globals. Or something even simpler, if all the methods you do in the connector will only be variants of `Get`, like `GetEvents` and `GetRegistration`, then you can add the `Get` to globals. It's not compulsory, but there if a situation arises. 

Additionally, in the SOAP library, there's another level, by the looks of it, for some SOAP services: see https://github.com/vpulim/node-soap#clientserviceportmethodargs-callback-options-extraheaders---call-a-method-using-a-specific-service-and-port
For this instance, you can simply use dot nation in the method name, and it'll work, because I'm using lodash's `_.get` to retrieve the relevant method https://github.com/trayio/threadneedle/blob/3f945b264421361a3e3a89b225de0ca3f013b164/lib/addMethod/addMethodSOAP.js#L161

Reply:
Nice - sounds good.